docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,319 @@
+```yaml
+sync:
+# include all files below src...
+- from: 'src/**/*'
+  to: app/
+# ...but do not consider js files for sync
+- from: 'src/**/*.js'
+  exclude: true
+```
+Also see #1766.
+
+Resolution: __Not Yet Resolved__
+   
+   
+#### Migration plan
+The new configuration supports all existing use-cases and is fully backwards compatible.
+An automatic schema upgrade shall be implemented.
+
+### Open Issues/Question
+
+**\<Overlapping copies\>** 
+Overlapping copies can be caused by at least two ways:
+
+1. The `COPY/ADD` instructions in the Dockerfile may copy to the same destination. For example:
+   ```dockerfile
+   COPY foo /foo
+   COPY bar/ /   # bar contains a file called "foo"
+   ```
+2. Two sync rules may copy files to the same destination. For example:
+   ```yaml
+   sync:
+   - from: foo
+     to: /baz
+   - from: bar
+     to: /baz
+   ```
+   
+Should such cases be treated as errors, warnings, or not be detected at all?
+
+Resolution: __Not Yet Resolved__
+
+**\<Transitive copies\>** 
+Multi-stage Dockerfiles can lead to quite contrived scenarios. For example:
+```dockerfile
+FROM scratch
+ADD foo baz
+FROM scratch
+COPY --from=0 baz bar
+```
+Ideally, the sync destination should be inferred as `foo -> bar`.
+
+This is quite an edge case and can easily be circumvented.
+However, it is quite complex to implement.
+Should we bother about this special case?
+   
+Resolution: __Not Yet Resolved__
+
+**\<Other builders\>**
+We don't know how to obtain a mapping of local paths to remote paths for other builders (jib, bazel).
+This may even have to be implemented upstream.
+Until we can support those builders, we to handle the case when a user tries to use inference with those builders.
+
+- Option 1: bail out with an error
+- Option 2: copy the file into the container with the destination set to the relative path wrt the context directory.
+
+Resolution: __Not Yet Resolved__
+
+**\<Upgrade hint\>**
+The path inference allows to simplify the sync specification considerably.
+Besides, subdirectory syncing makes the sync functionality even feasible for large projects that need to sync many folders into the container.
+Therefore, we could consider to advertise this new functionality when running Skaffold or when an old-style sync map as found.
+
+Resolution: __Not Yet Resolved__
+
+**\<SyncMap signature\>** 
+The new `SyncMap` interface provides some sync rules from the builder.
+However, the suggested signature is `func() map[string]string` with the convention to not flatten directories.
+To be more in line with the other sync functionality, it could also return `[]SyncRule`.
+Although being clearer, this has the downside of introducing a dependency from the builder package to the pipeline config package.
+If going that route, we should probably duplicate this struct in a proper place.
+However, is it necessary at all, or is the current signature good enough?
```

Comment:
I like the idea of having `SyncRule` however not sure how the implementation will look , like you mentioned.
Shd we consider this as an implementation detail and proceed?

Re: `SyncMap` maybe we should change the signature to`map[string][]string` to be in Sync with the `DependenciesForAtifacts` and also more flexible if same file is synced in multiple places.

Reply:
+1 let's punt this to implementation time - I think you can remove this from here - but I do like the idea of `SyncRule`.